src/modules/commander/mission_auth.cpp

```diff
@@ -0,0 +1,178 @@
+static orb_advert_t *mavlink_log_pub;
+static int command_ack_sub = -1;
+
+static uint8_t mission_auth_req_seq;
+
+static hrt_abstime auth_timeout;
+static hrt_abstime ack_timeout;
+
+static enum {
+	MISSION_AUTH_IDLE = 0,
+	MISSION_AUTH_WAITING_AUTH,
+	MISSION_AUTH_WAITING_AUTH_WITH_ACK,
+	MISSION_AUTH_MISSION_APPROVED
+} state = MISSION_AUTH_IDLE;
+
+static void mission_msg_auth_request_send(int16_t sequence, hrt_abstime now)
+{
+	struct vehicle_command_s cmd;
+
+	memset(&cmd, 0, sizeof(cmd));
+	cmd.command = MISSION_AUTH_REQ_COMMAND_ID;
+	cmd.param1 = sequence;
+	cmd.target_system = MISSION_AUTH_SYSTEM_ID;
+	cmd.target_component = 1;
+
+	if (handle_vehicle_command_pub == nullptr) {
+		handle_vehicle_command_pub = orb_advertise(ORB_ID(vehicle_command), &cmd);
+	} else {
+		orb_publish(ORB_ID(vehicle_command), handle_vehicle_command_pub, &cmd);
+	}
+
+	ack_timeout = now + MISSION_AUTH_CMD_ACK_TIMEOUT_USEC;
+}
+
+uint8_t mission_auth_check(const hrt_abstime now)
+{
+	switch (state) {
+	case MISSION_AUTH_IDLE:
+		/* no authentication in process? handle bellow */
+		break;
+	case MISSION_AUTH_MISSION_APPROVED:
+		return vehicle_command_ack_s::VEHICLE_RESULT_ACCEPTED;
+	case MISSION_AUTH_WAITING_AUTH:
+	case MISSION_AUTH_WAITING_AUTH_WITH_ACK:
+		return vehicle_command_ack_s::VEHICLE_RESULT_TEMPORARILY_REJECTED;
+	default:
+		return vehicle_command_ack_s::VEHICLE_RESULT_DENIED;
+	}
+
+	/* handling MISSION_AUTH_IDLE */
+	mission_auth_req_seq++;
+	mission_msg_auth_request_send(mission_auth_req_seq, now);
+	auth_timeout = now + MISSION_AUTH_CHECK_DEFAULT_TIMEOUT_USEC;
+	state = MISSION_AUTH_WAITING_AUTH;
+
+	mavlink_log_critical(mavlink_log_pub, "Requesting mission authorization...");
+
+	return vehicle_command_ack_s::VEHICLE_RESULT_TEMPORARILY_REJECTED;
+}
+
+void mission_auth_update(const hrt_abstime now)
+{
+	switch (state) {
+	case MISSION_AUTH_WAITING_AUTH:
+	case MISSION_AUTH_WAITING_AUTH_WITH_ACK:
+		/* handle bellow */
+		break;
+	case MISSION_AUTH_MISSION_APPROVED:
+		if (now > auth_timeout) {
+			state = MISSION_AUTH_IDLE;
+		}
+		return;
+	case MISSION_AUTH_IDLE:
+	default:
+		return;
+	}
+
+	/* handling MISSION_AUTH_WAITING_AUTH and MISSION_AUTH_WAITING_AUTH_WITH_ACK */
+	vehicle_command_ack_s command_ack;
+	bool updated = false;
+
+	command_ack.command = 0;
+
+	orb_check(command_ack_sub, &updated);
+	if (updated) {
+		orb_copy(ORB_ID(vehicle_command_ack), command_ack_sub, &command_ack);
+	}
+
+	if (command_ack.command == MISSION_AUTH_REQ_COMMAND_ID) {
+		switch (command_ack.result) {
+		case vehicle_command_ack_s::VEHICLE_RESULT_IN_PROGRESS:
+			state = MISSION_AUTH_WAITING_AUTH_WITH_ACK;
+			break;
+		case vehicle_command_ack_s::VEHICLE_RESULT_ACCEPTED:
+			mavlink_log_critical(mavlink_log_pub,
+					"Mission approved, allowing arm for the next %u seconds",
+					MISSION_APPROVED_TIMEOUT_USEC / 1000000);
+			state = MISSION_AUTH_MISSION_APPROVED;
+			auth_timeout = now + MISSION_APPROVED_TIMEOUT_USEC;
+			break;
```

Comment:
I have a suggestion here. From a user perspective, 2 arm commands are completely un-intuitive. I would suggest that you request authorization on pre-arm (safety switch pressed and system de-saftied).

Then wait for arming from the operator. If the auth has succeeded by then, directly arm - otherwise refuse arming citing no authorization yet. How does that sound?